blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,292 @@
+
+At the time of writing this post, these types are documented under Typed Racket's [experimental features](https://docs.racket-lang.org/ts-reference/Experimental_Features.html). They are experimental in the sense that they are relatively new. We look forward to improving their usefulness by incorporating user feedback.
+
+In the near future, if you wish to experiment with refinement and/or dependent function types you may want to grab a [snapshot build](https://pre.racket-lang.org/installers/), as it will have additional features and/or bug fixes from user feedback.
+
+
+## Refinement Types
+
+Let us examine the type of `vector-ref` in Typed Racket:
+
+```racket
+(All (a) (-> (Vectorof a) Integer a))
+```
+
+It states that the index (the second argument) must be an `Integer`, i.e. any whole-number value in the exclusive range (-∞ , ∞). This description, however, is somewhat coarse: we obviously cannot use _any_ integer. To avoid a runtime error when indexing into a vector `v`, we must use an integer in the range [0, `(vector-length v)`). With a refinement type we can precisely describe this subset of the integers:
+
+```racket
+(Refine [i : Integer] (<= 0 i (- (vector-length v) 1)))
+```
+
+This refinement type describes values `i` of type `Integer` such that `(<= 0 i (- (vector-length v) 1))` produces `#true`. The documentation for [Refine](http://docs.racket-lang.org/search/index.html?q=Refine) describes precisely what propositions can be used in refinements. Today it includes logical statements about the types of terms, standard logical combinators (e.g. and, or), and linear integer constraints.
+
+## Dependent function types
+
+By combining refinement types with dependent function types (i.e. functions whose domain and range may depend on argument and/or program values), we can now give a more precise function type for many basic vector operations, e.g.:
+
+```racket
+(: safe-vector-ref 
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : (v) (Refine [i : Integer]
+                                  (<= 0 i (- (vector-length v) 1)))])
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Here we have stated that `safe-vector-ref`'s second argument (`n`) depends on its first argument (`v`) and that the type of `n` is an `Integer` which is a valid index for `v`.
+
+If we try and use `safe-vector-ref` with a vector and index which Typed Racket cannot prove satisfy the respective types:
+
+```racket
+(safe-vector-ref (vector 0 1) 2)
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+When a refinement on an argument's type may also be written conveniently as a "precondition", e.g.:
```

Comment:
This is a pretty big and unexpected jump.  Should the reader know about `#:pre` and the idea that a refinement may also be formulated as a precondition (I didn't)?  Also, you should explain the precondition first, then the error message, not the reverse.

One thing that's not really addressed later in the document is when should someone use a refinement vs precondition.